torch/csrc/distributed/c10d/FlightRecorder.cpp

```diff
@@ -253,65 +650,22 @@ std::optional<size_t> FlightRecorder::record(
     Event* end,
     std::chrono::milliseconds timeout_ms,
     std::shared_ptr<ProcessGroupStatus> pg_status,
-    bool isP2P) {
-  if (!enabled_) {
-    return std::nullopt;
-  }
-  if (all_pg_status_.find(pg_id) == all_pg_status_.end()) {
-    // Current pg_status is not in FR.
-    all_pg_status_[pg_id] = std::move(pg_status);
-  }
-  auto traceback =
-      torch::CapturedTraceback::gather(true, true, capture_cpp_stack_);
-  std::lock_guard<std::mutex> guard(mutex_);
-
-  auto te = Entry{
-      id_,
-      pg_id,
-      pg_name,
-      collective_seq_id,
-      p2p_seq_id,
-      op_id,
-      std::move(profiling_name),
-      std::move(traceback),
-      start,
-      end,
-      c10::getTime(),
-      timeout_ms.count(),
-      isP2P,
-      std::nullopt,
-      std::nullopt,
-      std::nullopt,
-      {},
-      {},
-      {},
-      {},
-      {},
-      false};
-
-  for (const auto& input : inputs) {
-    c10::IntArrayRef sizes = input.sizes();
-    te.input_dtypes_.push_back(input.dtype().toScalarType());
-    te.input_dims_.push_back(static_cast<int64_t>(sizes.size()));
-    te.sizes_.insert(te.sizes_.end(), sizes.begin(), sizes.end());
-  }
-
-  for (const auto& output : outputs) {
-    c10::IntArrayRef sizes = output.sizes();
-    te.output_dtypes_.push_back(output.dtype().toScalarType());
-    te.output_dims_.push_back(static_cast<int64_t>(sizes.size()));
-    te.sizes_.insert(te.sizes_.end(), sizes.begin(), sizes.end());
-  }
-
-  if (entries_.size() < max_entries_) {
-    entries_.emplace_back(std::move(te));
-  } else {
-    entries_[next_++] = std::move(te);
-    if (next_ == max_entries_) {
-      next_ = 0;
-    }
-  }
-  return id_++;
+    bool isP2P) {
+  return recordImpl<FlightRecorder, FlightRecorder::Entry, Event>(
```

Comment:
So you moved the implementation out into an independent function, calling a need to template on all classes?

Would it be easier if we keep the implementation in the class method as before, and only template the method?
```
template< typename EventType>
FlightRecorder::record(EventType* start, ...)

// At call site
FlightRecorder* fr;
fr->template record<CUDAEvent>(start, ...);  // template keyword needed here
```

Reply:
we can but the thing is that we will use that event to initiate the entry inside (which also takes in an event)..